core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataMappingService.java

```diff
@@ -231,10 +230,15 @@ public void refreshMapping(final String index, final String indexUUID, final Str
                     for (String index : request.indices()) {
                         if (currentState.metaData().hasIndex(index)) {
                             // if we don't have the index, we will throw exceptions later;
-                            if (indicesService.hasIndex(index) == false) {
+                            if (indicesService.hasIndex(index) == false || indicesToClose.contains(index)) {
                                 final IndexMetaData indexMetaData = currentState.metaData().index(index);
-                                IndexService indexService = indicesService.createIndex(nodeServicesProvider, indexMetaData, Collections.EMPTY_LIST);
-                                indicesToClose.add(indexMetaData.getIndex());
+                                IndexService indexService;
+                                if (indicesService.hasIndex(index) == false) {
+                                    indexService = indicesService.createIndex(nodeServicesProvider, indexMetaData, Collections.EMPTY_LIST);
+                                    indicesToClose.add(index);
+                                } else {
+                                    indexService = indicesService.indexService(index);
+                                }
                                 // make sure to add custom default mapping if exists
                                 if (indexMetaData.getMappings().containsKey(MapperService.DEFAULT_MAPPING)) {
                                     indexService.mapperService().merge(MapperService.DEFAULT_MAPPING, indexMetaData.getMappings().get(MapperService.DEFAULT_MAPPING).source(), false, request.updateAllTypes());
```

Comment:
Re the addition of types - I think we can also just check if it's not there? I'm thinking about batching multiple requests for the same type ( commenting here because the code in question is not in the scope of the change)